File: data/preprocessing_codes/compute_graph_kernels.m



function [] = compute_graph_kernel(kernel_type)

  addpath './graphkernels'
  addpath './graphkernels/labeled'
  addpath './graphkernels/unlabeled'
  addpath './graphkernels/svm'
  addpath 'graphkernels/unlabeled/allgraphlets/'
  addpath 'graphkernels/unlabeled/samplinggraphlets/'
  addpath 'graphkernels/unlabeled/connectedgraphlets/'

  labels = dlmread('../DTPNCI2015/results/ncicancer_labels');
  labels = labels(1:5000);

  X = [];
  for i = 1:5000
    load(sprintf('../structures/MATLABfiles/%d.mat', labels(i)));
    X = [X,N];
  end

  if strcmp(kernel_type,'lRWkernel')
    [K,t] = lRWkernel(X, 0.005, 1);
    K = normalizekm(K);
    dlmwrite(sprintf('../DTPNCI2015/results/ncicancer_kernel_graph_%s',kernel_type), K);
    
  elseif strcmp(kernel_type,'WL')
    [K,t] = WL(X,1,1);
    K = normalizekm(K{1});
    dlmwrite(sprintf('../DTPNCI2015/results/ncicancer_kernel_graph_%s',kernel_type), K);

  elseif strcmp(kernel_type,'WLedge')
    [K,t] = WLedge(X,1,1);
    K = normalizekm(K{1});
    dlmwrite(sprintf('../DTPNCI2015/results/ncicancer_kernel_graph_%s',kernel_type), K);

  elseif strcmp(kernel_type,'WLspdelta')
    [K,t] = WLspdelta(X,1,1,0);
    K = normalizekm(K{1});
    dlmwrite(sprintf('../DTPNCI2015/results/ncicancer_kernel_graph_%s',kernel_type), K);

  elseif strcmp(kernel_type,'RGkernel1')
    [K,t] = RGkernel(X,1);
    K = normalizekm(K);
    dlmwrite(sprintf('../DTPNCI2015/results/ncicancer_kernel_graph_%s',kernel_type), K);

  elseif strcmp(kernel_type,'l3graphletkernel')
    [K,t] = l3graphletkernel(X);
    K = normalizekm(K);
    dlmwrite(sprintf('../DTPNCI2015/results/ncicancer_kernel_graph_%s',kernel_type), K);

  elseif strcmp(kernel_type,'untilpRWkernel4')
    [K,t] = untilpRWkernel(X,4,1);
    K = normalizekm(K);
    dlmwrite(sprintf('../DTPNCI2015/results/ncicancer_kernel_graph_%s',kernel_type), K);

  elseif strcmp(kernel_type,'untilpRWkernel6')
    [K,t] = untilpRWkernel(X,6,1);
    K = normalizekm(K);
    dlmwrite(sprintf('../DTPNCI2015/results/ncicancer_kernel_graph_%s',kernel_type), K);

  elseif strcmp(kernel_type,'untilpRWkernel8')
    [K,t] = untilpRWkernel(X,8,1);
    K = normalizekm(K);
    dlmwrite(sprintf('../DTPNCI2015/results/ncicancer_kernel_graph_%s',kernel_type), K);

  elseif strcmp(kernel_type,'untilpRWkernel10')
    [K,t] = untilpRWkernel(X,10,1);
    K = normalizekm(K);
    dlmwrite(sprintf('../DTPNCI2015/results/ncicancer_kernel_graph_%s',kernel_type), K);

  elseif strcmp(kernel_type,'spkernel')
    [K,t] = spkernel(X);
    K = normalizekm(K);
    dlmwrite(sprintf('../DTPNCI2015/results/ncicancer_kernel_graph_%s',kernel_type), K);

  elseif strcmp(kernel_type,'SPkernel')
    [K,t] = SPkernel(X);
    K = normalizekm(K);
    dlmwrite(sprintf('../DTPNCI2015/results/ncicancer_kernel_graph_%s',kernel_type), K);

  elseif strcmp(kernel_type,'RWkernel')
    [K,t] = RWkernel(X,0.005);
    K = normalizekm(K);
    dlmwrite(sprintf('../DTPNCI2015/results/ncicancer_kernel_graph_%s',kernel_type), K);

  elseif strcmp(kernel_type,'gestkernel3')
    [K,t] = gestkernel3(X,100);
    K = normalizekm(K);
    dlmwrite(sprintf('../DTPNCI2015/results/ncicancer_kernel_graph_%s',kernel_type), K);

  elseif strcmp(kernel_type,'gestkernel4')
    [K,t] = gestkernel4(X,100);
    K = normalizekm(K);
    dlmwrite(sprintf('../DTPNCI2015/results/ncicancer_kernel_graph_%s',kernel_type), K);

  elseif strcmp(kernel_type,'connectedkernel3')
    [K,t] = connectedkernel(X,3);
    K = normalizekm(K);
    dlmwrite(sprintf('../DTPNCI2015/results/ncicancer_kernel_graph_%s',kernel_type), K);

  elseif strcmp(kernel_type,'connectedkernel4')
    [K,t] = connectedkernel(X,4);
    K = normalizekm(K);
    dlmwrite(sprintf('../DTPNCI2015/results/ncicancer_kernel_graph_%s',kernel_type), K);

  elseif strcmp(kernel_type,'connectedkernel5')
    [K,t] = connectedkernel(X,5);
    K = normalizekm(K);
    dlmwrite(sprintf('../DTPNCI2015/results/ncicancer_kernel_graph_%s',kernel_type), K);

  end



exit;
end

File: data/preprocessing_codes/compute_graph_kernels.py


#!/usr/bin/env python
#-*- coding: iso-8859-15 -*-


# use packages
import sys
import os
import math
import re
import Queue
import time
import optparse
import random
import commands
import logging
from threading import ThreadError
from threading import Thread
sys.path.append('/cs/taatto/group/urenzyme/workspace/netscripts/')
from get_free_nodes import get_free_nodes


job_queue = Queue.Queue()

# worker class
# job is a tuple of parameters
class Worker(Thread):
  def __init__(self, job_queue, node):
    Thread.__init__(self)
    self.job_queue  = job_queue
    self.node = node
  def run(self):
    all_done = 0
    while not all_done:
      try:
        job = self.job_queue.get(0)
        time.sleep(random.randint(5000,6000) / 1000.0)	# sleep random time
        single_thread(self.node, job)
      except Queue.Empty:
        all_done = 1
  pass

def single_thread(node,job):
  (job_id,job_content) = job
  logging.info(" %s->%s:%s" % (node, job_id, job_content))
  try:
    singleres = commands.getoutput(''' ssh %s " cd /cs/taatto/group/urenzyme/workspace/molecular_classification/data/preprocessing_codes; nohup matlab -nodisplay -nodesktop -nojvm -r 'compute_graph_kernels %s'  "  ''' % (node,job_content))
    logging.info(" %s-|%s" % (node, job_id))
  except:
    logging.info(" %s-<%s" % (node, job_id))
    job_queue.put((job))
  pass


def compute_graph_kernels_in_parallel():
  cluster = get_free_nodes()[0]
  kernelnamelist =  ['lRWkernel','WL',  'WLedge',  'WLspdelta',  'RGkernel1',  'l3graphletkernel',  'untilpRWkernel4',  'untilpRWkernel6',  'untilpRWkernel8',  'untilpRWkernel10',  'spkernel',  'SPkernel',  'RWkernel',  'gestkernel3',  'gestkernel4',  'connectedkernel3',  'connectedkernel4',  'connectedkernel5']
  # generate jobs and so on
  job_id = 0
  job_size = 1
  for kernelname in kernelnamelist:
    kernelresultfile = "../DTPNCI2015/results/ncicancer_kernel_graph_%s" % (kernelname)
    if os.path.exists(kernelresultfile):
      continue
    job_id = job_id + 1
    job_content = kernelname
    job_queue.put((job_id, job_content))
  logging.info("Processing jobs ...")
  # processing jobs
  job_size = job_queue.qsize()
  logging.info("In total %d jobs" % job_size)
  loadpernode = 1 
  threads = []
  counter = 0
  for node in cluster:
    for i in range(loadpernode):
      t = Worker(job_queue, node)
      time.sleep(0.5)
      counter = counter +1
      if counter > job_size:
        break
      try:
        t.start()
        threads.append(t)
      except ThreadError:
        logging.warning("\t\tThread error!")
  for t in threads:
    t.join()
  pass




compute_graph_kernels_in_parallel()







